### PR TITLE
gcoap: Do not send responses from multicast addresses

### PR DESCRIPTION
**Pull request by chrysn**
### Contribution description

Since https://github.com/RIOT-OS/RIOT/pull/18026, CoAP requests to multicast addresses (eg. `ff02::1`) came back from that exact address, which Linux rightfully just drops.

The fix uses the existing multicast check from https://github.com/RIOT-OS/RIOT/pull/17978 (thanks @benpicco for making me write this as dedicated function, I just had to generalize it removing one struct layer), and foregoes setting the source address when responding to multicasts.

### Testing procedure

* Run the gcoap example
* Send a CoAP request to a multicast address RIOT listens to, eg. `./aiocoap-client coap://'[ff02::1%tapbr0]'/.well-known/core --non`

Before, this got no response (while you see it arrive on wireshark). After, you get a correct response with two lines of note:

```
WARNING:coap:Sending request to multicast via unicast request method
Response arrived from different address; base URI is coap://[fe80::3c63:beff:fe85:ca96%tapbr0]/.well-known/core
```

(The former is aiocoap telling us that we're not using the nonexistent multicast API so it's really more of an anycast, the latter is useful factual information).

### CODE REVIEW

**Review by benpicco — APPROVED** [review]

**Comment by chrysn** [comment]
Thanks

bors merge

File: sys/net/application_layer/gcoap/gcoap.c [hunk 1]
```diff
@@ -345,18 +346,25 @@ static void _on_sock_udp_evt(sock_udp_t *sock, sock_async_flags_t type, void *ar
...
-        /* make sure we reply with the same address that the request was destined for */
+        /* make sure we reply with the same address that the request was
+         * destined for -- except in the multicast case */
+        sock_udp_aux_tx_t *aux_out_ptr;
         sock_udp_aux_tx_t aux_out = {
             .flags = SOCK_AUX_SET_LOCAL,
             .local = aux_in.local,
         };
+        if (_ep_is_multicast(&aux_in.local)) {
+            aux_out_ptr = NULL;
```
Comment (OlegHahm) [inline sys/net/application_layer/gcoap/gcoap.c:357]: Probably I'm too late, but a comment about the semantics of setting this pointer to NULL here would have been nice.

Comment (chrysn) [reply sys/net/application_layer/gcoap/gcoap.c:359]: Good point, given it takes several indirections before this wins up in a function where one can read the documentation.

Comment added, let's see how mad bors will get at me for throwing logs before its merge train.

**Comment by chrysn** [comment]
bors merge

**Comment by benpicco** [comment]
bors merge

**Pull request merged**